src/expr/src/relation/func.rs

```diff
@@ -513,6 +513,95 @@ impl AggregateFunc {
+    pub fn outer_agg(&self) -> Option<(AggregateFunc, Option<crate::UnaryFunc>)> {
+        match self {
+            AggregateFunc::Count => Some((
+                AggregateFunc::SumInt64,
+                Some(crate::UnaryFunc::CastDecimalToInt64(0)),
+            )),
+            AggregateFunc::SumInt32 => Some((
+                AggregateFunc::SumInt64,
+                Some(crate::UnaryFunc::CastDecimalToInt64(0)),
+            )),
+            AggregateFunc::SumInt64 => Some((AggregateFunc::SumDecimal, None)),
+            // we don't yet have an aggregation to flat map jsonb lists into bigger lists
+            AggregateFunc::JsonbAgg => None,
+            _ => Some((self.clone(), None)),
+        }
+    }
+
+    /// Whether an aggregate always returns the same answer when evaluated on a
+    /// collection as when evaluated on the same collection but with duplicate
+    /// values removed.
+    ///
+    /// Suppose we have a collection `A=[1, 1, 2, 4, 4]`. The maximum value in
+    /// the collection does not change if we remove duplicate values. However,
+    /// the sum of the values in the collection will.
+    pub fn ignores_duplicates(&self) -> bool {
+        !matches!(
+            self,
+            AggregateFunc::MaxInt32
+                | AggregateFunc::MaxInt64
+                | AggregateFunc::MaxFloat32
+                | AggregateFunc::MaxFloat64
+                | AggregateFunc::MaxDecimal
+                | AggregateFunc::MaxBool
+                | AggregateFunc::MaxString
+                | AggregateFunc::MaxDate
+                | AggregateFunc::MaxTimestamp
+                | AggregateFunc::MaxTimestampTz
+                | AggregateFunc::MinInt32
+                | AggregateFunc::MinInt64
+                | AggregateFunc::MinFloat32
+                | AggregateFunc::MinFloat64
+                | AggregateFunc::MinDecimal
+                | AggregateFunc::MinBool
+                | AggregateFunc::MinString
+                | AggregateFunc::MinDate
+                | AggregateFunc::MinTimestamp
+                | AggregateFunc::MinTimestampTz
+                | AggregateFunc::Dummy
+        )
+    }
+
+    /// Whether the result applying the aggregate to a literal is
+    /// dependent on the specific number of rows in the input.
+    ///
+    /// The result of an aggregate on a literal is usually dependent on
+    /// whether the input is empty or not. For example, `max('constant')`
+    /// will return `null` if the input is empty and `'constant'` otherwise.
+    ///
+    /// But the results of some aggregates on a literal further depend on the
+    /// specific number of rows in the input. For example, `count('constant')`
+    /// will return the number of rows in the input, and `sum(2)` will return
+    /// two times the number of rows in the input.
+    ///
+    /// This function will return `true` for aggregates like `count`
+    /// and `false` for an aggregate like `max`.
```

Comment:
What should it do for other aggregations like `jsonb_agg`?

Reply:
Returns false because just like `max`, if the input argument is an argument, it returns null if the table is empty and the constant if the table is not empty.

output of `jsonb_agg` as run in `psql`:
```
andiwang=# create table blah (a int, b int);
CREATE TABLE
andiwang=# select jsonb_agg('{"a": 42, "b":53}'::jsonb) from blah;
 jsonb_agg 
-----------
 
(1 row)

andiwang=# insert into blah values(1, 2);
INSERT 0 1
andiwang=# select jsonb_agg('{"a": 42, "b":53}'::jsonb) from blah;
```

Reply:
Ok, maybe this simplifies when the comment above is shaken out, that this behavior is something sorted out at the `Hir` level and becomes irrelevant at the `Mir` level. The comment was mainly because two cases were given as examples, and they did not cover the space of aggregations we support.

Reply:
Let me try to explain what I'm trying to do. Suppose we have a query,
```
select count(true) from foo, bar where foo.col1 = bar.col1
```
I feel it is reasonable to push down count(true) to either input, to make it look like
```
select sum(sub_count)
from (select foo.col1, count(true) as sub_count from foo) foo_reduced, bar 
where foo_reduced.col1 = bar.col1
```
or
```
select sum(sub_count) 
from foo, (select bar.col1, count(true) as sub_count from bar) bar_reduced
where foo.col1 = bar_reduced.col1
```

------

On the other hand, if we had 
```
select max(true) from foo, bar where foo.col1 = bar.col1
```
to me, it feels silly to push down a reduction in this case to make it look like
```
select max(max_true) 
from (select foo.col1, max(true) as max_true from foo group by foo.col1) as foo_reduced, bar
where foo_reduced.col1 = bar.col1
```

I think a better transform would be if the above query became:
```
select true from foo, bar where foo.col1 = bar.col1
```
or
```
select true from foo, (select distinct col1 from bar) distinct_bar where foo.col1 = distinct_bar.col1
```
But I consider this better transformation to be out of the scope of the PR, so I just wanted a function that allows me to separate the aggregations on a constant that are worth pushing down vs. the ones that are not.